main/src/cgeo/geocaching/ui/CalculatorVariable.java

```diff
@@ -0,0 +1,304 @@
+        cacheDirty = theCacheState;
+    }
+
+    // Method to returned the cached values as a String
+    // This is used to display the result in the UI.
+    public String getCachedString() {
+        final String rv;
+
+        if (variableData.expression == null || variableData.expression.length() == 0)
+            rv = "_";
+        else if (Double.isNaN(getCachedValue()))
+            rv = "*";
+        else
+            rv = "" + (int) getCachedValue();
+
+        return rv;
+    }
+
+    // Method to returned the cached values as a Double
+    // This is used to compute the value of another variable.
+    public double getCachedValue() {
+        return cachedValue;
+    }
+
+    public void setCachedValue(final double cachedValue) {
+        this.cachedValue = cachedValue;
+        setCacheDirty(false);
+
+        // Make the name colour grey if value is invalid
+        final int nameColour = Double.isNaN(getCachedValue()) ? ContextCompat.getColor(getContext(), R.color.text_hint_dark) : Color.WHITE;
+        name.setTextColor(nameColour);
+    }
+
+    public char getName() {
+        return variableData.name;
+    }
+
+    public String getExpression() {
+        return variableData.expression;
+    }
+
+    public void setExpressionHint(final CharSequence theHint) {
+        expression.setHint(theHint);
+    }
+
+    public int getExpressionId() {
+        return expression.getId();
+    }
+
+    public void setExpressionId(final int id) {
+        expression.setId(id);
+    }
+
+    public String EvaluateString(final List<CalculatorVariable> dependantVariables) {
+        if (isCacheDirty())
+            EvaluateDouble(dependantVariables);
+
+        return getCachedString();
+    }
+
+    private double EvaluateDouble(final List<CalculatorVariable> dependantVariables) {
+        if (isCacheDirty()) {
+            String expression = getExpression();
+
+            if (dependantVariables != null) {
+                for (final CalculatorVariable depVar : dependantVariables)
+                    expression = expression.replace("" + depVar.getName(), "(" + depVar.EvaluateDouble(null) + ")");
+            }
+
+            try {
+                setCachedValue(eval(expression));
+            } catch (final Exception e) {
+                setCachedValue(Double.NaN);
+            }
+        }
+
+        return getCachedValue();
+    }
+
+    // This simple evaluation algorithm was created by user 'Boann' and released to the public domain on Stack Overflow:
+    // https://stackoverflow.com/questions/3422673/evaluating-a-math-expression-given-in-string-form
+    // I added the '%' operator myself.
+    // Note that the 'sqrt', 'cos', 'sin' and 'tan' functions don't work in this implementation as
+    // expected as the individual letters are automatically replaced with their corresponding
+    // expressions as they are interpreted as being individual variables.
+    private static double eval(final String str) {
```

Comment:
I think this method would be better split into a new `cgeo.geocaching.utils` class (`CalculatorUtils` maybe) with no Android dependency, that would make it more easily testable without requiring an Android device or emulator.